### DISCUSSION TITLE
SyntaxError: "undefined" is not valid JSON

### DISCUSSION THREAD

**Discussion by danimadmolil** [vercel/next.js#48651] [Help] [answered]
### Summary

i currently create a new nextjs application;

here is counter component
```
"use client";

import { useState } from "react";

export default function Counter() {
  const [count, setCount] = useState(0);

  return (
    <div>
      <p>You clicked {count} times</p>
      <button onClick={() => setCount(count + 1)}>Click me</button>
    </div>
  );
}
```
here is the root page

```
"use client";
import Image from "next/image";
import { Inter } from "next/font/google";
import Counter from "../components/Counter";

const inter = Inter({ subsets: ["latin"] });

export default function Home() {
  return (
    <main className="p-3">
      <Counter/>
    </main>
  );
}
```

i think problem is that using 'use client'; because if i remove hooks and the 'use client' flag in header component the error will go away.
any body know why is tha? 
thanks

**Comment by liamtran96** [comment]
Yes I have the same problem. I am using window and  next 13.2.0 version. But the issue does not display on my mac

**Reply by theviveksahu** [reply]
I am running next 13.2.4 and can confirm issue is in mac as well

**Comment by danimadmolil** [comment, marked as answer]
I change all the pages and components and layouts to arrow functions and it resolve the issue.

**Reply by mamlzy** [reply]
it works thanks!
changed from `export default Component() {}` to `const Component = () => {};
export default Component;`

**Reply by HialeahLiam** [reply]
What a weird bug haha. Solved it for me too on a windows laptop!

I think this is a bug, right?

**Reply by danimadmolil** [reply]
yes it should be a bug, version 13 is not stable yet.

**Reply by theviveksahu** [reply]
Hit this bug yesterday on macOS Catalina. Tried with and without "use client". All my components are already arrow functions, not sure what to do :(

**Reply by nisbusra** [reply]
> Hit this bug yesterday on macOS Catalina. Tried with and without "use client". All my components are already arrow functions, not sure what to do :(

I have the same problem as you. Changed them all but still, bugs are there. Any update?

**Comment by psygo** [comment]
Still happening here

**Comment by marko-hologram** [comment]
I had this happen, but it wasn't related to arrow functions or the way components were really rendered. It was related to cache, fetch and returning `undefined` from a cache function when rendering a server component page that had dynamic metadata.

So my code was something like this:

```tsx
import { unstable_cache as cache } from "next/cache";
import { redirect } from "next/navigation";

const fetchUser = cache(
  async (userId: string) => {
    const response = await fetch(`https://example.com/api/users/${userId}`, {
      next: {
        // Keep it in cache for 10 minutes
        revalidate: 600,
      },
    });

    if (!response.ok) {
      // This return was the issue. It should be `return null` instead of `return undefined`
      return undefined;
    }

    return response.json();
  },
  ["single-user-key"],
  {
    // Keep it in cache for 10 minutes
    revalidate: 600,
  }
);

export default async function UserSinglePage({ params }: { params: { userId: string } }) {
  const userData = await fetchUser(params.userId);

  if (!userData) {
    redirect("/some-page");
  }

  return <WhateverComponent userData={userData} />;
}

export async function generateMetadata({ params }: { params: { userId: string } }) {
  const userData = await fetchUser(params.userId);

  if (!userData) {
    return {};
  }

  return {
    title: userData.name,
    ...
  };
}
```

Problem seemed to be that cached data gets stringified somehow, I'm guessing with `JSON.stringify()`. Then when this data was being retrieved at some point it was retrieved using `JSON.parse()`. The problem here is parsing undefined because that produces an error. Parsing null actually produces `null` value and doesn't error.

```ts
JSON.parse(undefined); // This produces an error
JSON.parse(null); // This outputs `null`
```

Firefox produces this error "Uncaught SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data"
Chrome produces this error "Uncaught SyntaxError: "undefined" is not valid JSON"

I managed to resolve this by just returning null from cached function so that it can be parsed properly and not produce a runtime error.

---

Seeing solution mention "**_I change all the pages and components and layouts to arrow functions and it resolve the issue._**" is weird. I think that was either a bug or just the app being rebuilt because of such a change busted some broken cache and things started working. Since Next.js caches a lot of stuff heavily, maybe the issue was that somewhere some component was returning undefined instead of JSX or null and that stored an invalid entry in cache and caused issues later when cache was being retrieved. I'm just speculating because it's hard to say.

Hopefully this can help someone else if they run into this issue as I did.

**Reply by travisluong** [reply]
Thanks. Was using `unstable_cache` for my db queries. Removing `unstable_cache` seems to have fixed the problem for me.

**Reply by marko-hologram** [reply]
@travisluong But removing `unstable_cache` then removes caching. Are you sure that is what you want to do?

**Comment by professor-oats** [comment]
Hello! Nice to see a fresh update on this topic.
I have a similar issue that I run into.

I can't seem to make the deploy run properly. The logs give error 500 when using GET req on '/' - but instead of 'undefined', as in TS, I get "SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at JSON.parse (<anonymous>) 
    at parseJSONFromBytes (node:internal/deps/undici/undici:5329:19)"
    
 I have tried as a workaround to have "app.use(bodyParser.json());" to ensure that the JSON would be able to handle parse of HTML as well.
 Currently I feel little leftout from the debugging options of Vercel, since it doesn't really connect to what point in the src or code that the GET encounter this error.
 
 Builds completely without errors and I believe my environment variables may be correct since favicon can be loaded from where I have the root '/'.
 
 Similar of TS but another case.
 Same error as above for the '/api/users/me/' that is fetched on as in:
 
 "const meRes = await fetch(`${process.env.NEXT_PUBLIC_SERVER_URL}/api/users/me`, "
 
 Any ideas how to address this?
 Thanks.
 
 Have yet to try to convert everything to arrow functions, but boy I would love to find another path of testing.
